### PR TITLE
TRD: QC: Write corrected charges

### PR DESCRIPTION
**Pull request by f3sch**
TrackingQC produces now also the corrected charges (for now baring gain calibration) including z-row merging

Tagging @jcrkovsk for approval, as this pulls the krypton calibrations from the ccdb and corrects (hopefully) the charges by multiplication of the local gain factor (correct?).

### CODE REVIEW

File: Detectors/TRD/base/include/TRDBase/PadParameters.h [hunk 1]
```diff
@@ -74,10 +74,10 @@ int PadParameters<T>::getChannel(int col, int row) const
     return row + col * constants::NROWC0;
   } else if (mData.size() == constants::NROWC1 * constants::NCOLUMN) {
     return row + col * constants::NROWC1;
-  } else {
-    LOG(error) << "Wrong number of channels set: " << mData.size();
   }
-  return -1;
+
+  LOG(error) << "Wrong number of channels set: " << mData.size();
```
Comment (martenole) [inline Detectors/TRD/base/include/TRDBase/PadParameters.h]: Why the change here? In case the class is not initialized you would now get two different error messages

File: Detectors/TRD/qc/src/Tracking.cxx [hunk 2]
```diff
@@ -117,19 +124,25 @@ void Tracking::checkTrack(const TrackTRD& trkTrd, bool isTPCTRD)
...
-      mTrackletsRaw[trkltId].getQ2(),
+      tracklet.getQ0(),
+      tracklet.getQ1(),
+      tracklet.getQ2(),
+    };
+    auto correction = mLocalGain.getValue(tracklet.getHCID() / 2, tracklet.getPadCol(), tracklet.getPadRow());
+    qcStruct.trackletCorCharges[iLayer] = {
+      static_cast<float>(tracklet.getQ0()) * correction,
+      static_cast<float>(tracklet.getQ1()) * correction,
+      static_cast<float>(tracklet.getQ2()) * correction,
```
Comment (martenole) [inline Detectors/TRD/qc/src/Tracking.cxx]: don't think the `static_cast`s are needed here, or?

File: Detectors/TRD/qc/src/Tracking.cxx [hunk 3]
```diff
@@ -37,6 +38,11 @@ void Tracking::setInput(const o2::globaltracking::RecoContainer& input)
   mTracksTPCTRD = input.getTPCTRDTracks<TrackTRD>();
   mTrackletsRaw = input.getTRDTracklets();
   mTrackletsCalib = input.getTRDCalibratedTracklets();
+
+  if (!mCCDBInit) {
+    mCCDBInit = true;
+    mLocalGain = *(pc.inputs().get<o2::trd::LocalGainFactor*>("localgainfactors"));
```
Comment (martenole) [inline Detectors/TRD/qc/src/Tracking.cxx]: I would prefer to set the CCDB object from the TRDGlobalTrackingQCSpec, e.g. via a new method `Tracking::setLocalGainFactors()`. This could be called e.g. from inside `TRDGlobalTrackingQC::updateTimeDependentParams()`. Then you don't need the init flag and the additional includes

File: Detectors/TRD/base/include/TRDBase/PadParameters.h (hunk 1 above)
Comment (f3sch) [reply Detectors/TRD/base/include/TRDBase/PadParameters.h]: Right missed the missing return in the if

File: Detectors/TRD/qc/src/Tracking.cxx (hunk 2 above)
Comment (f3sch) [reply Detectors/TRD/qc/src/Tracking.cxx]: probably not

File: Detectors/TRD/qc/include/TRDQC/Tracking.h [hunk 4]
```diff
@@ -83,12 +86,14 @@ struct TrackQC {
...
-  std::array<int, constants::NLAYER> trackletMcm{};                                      ///< the MCM number of the tracklet
-  std::array<float, constants::NLAYER> trackletChi2{};                                   ///< estimated chi2 for the update of the track with the given tracklet
-  std::array<std::array<int, constants::NCHARGES>, constants::NLAYER> trackletCharges{}; ///< charges of tracklets
-  ClassDefNV(TrackQC, 4);
+  std::array<int, constants::NLAYER> trackletHCId{};                                          ///< the half-chamber ID of the tracklet
+  std::array<int, constants::NLAYER> trackletRob{};                                           ///< the ROB number of the tracklet
+  std::array<int, constants::NLAYER> trackletMcm{};                                           ///< the MCM number of the tracklet
+  std::array<float, constants::NLAYER> trackletChi2{};                                        ///< estimated chi2 for the update of the track with the given tracklet
+  std::array<std::array<int, constants::NCHARGES>, constants::NLAYER> trackletCharges{};      ///< charges of tracklets
```
Comment (bazinski) [inline Detectors/TRD/qc/include/TRDQC/Tracking.h:91]: Has the whole rule of not doing arrays of arrays been dropped?

File: Detectors/TRD/qc/src/Tracking.cxx (hunk 3 above)
Comment (f3sch) [reply Detectors/TRD/qc/src/Tracking.cxx]: Thanks for the review. Please check if these changes work for you.

File: Detectors/TRD/qc/include/TRDQC/Tracking.h (hunk 4 above)
Comment (tdietel) [reply Detectors/TRD/qc/include/TRDQC/Tracking.h:91]: I thought the problem were variable-size containers, e.g. vectors. Arrays as fixed size containers should be ok, and I think our digits actually contain an array.

File: Detectors/TRD/qc/include/TRDQC/Tracking.h [hunk 5]
```diff
@@ -118,13 +123,21 @@ class Tracking
   // Make output accessible to DPL processor
   std::vector<TrackQC>& getTrackQC() { return mTrackQC; }
 
+  // Set the local gain factors with values from the ccdb
+  void setLocalGainFactors(o2::framework::ProcessingContext& pc)
```
Comment (martenole) [inline Detectors/TRD/qc/include/TRDQC/Tracking.h]:
```suggestion
  void setLocalGainFactors(const o2::trd::LocalGainFactor* padGain)
```
Then you can drop the Framework includes above

File: Detectors/TRD/base/include/TRDBase/PadParameters.h [hunk 6]
```diff
@@ -77,7 +77,8 @@ int PadParameters<T>::getChannel(int col, int row) const
   } else {
     LOG(error) << "Wrong number of channels set: " << mData.size();
   }
-  return -1;
+
+  return 0;
```
Comment (martenole) [inline Detectors/TRD/base/include/TRDBase/PadParameters.h]: I would just leave this as is, or why do you prefer a valid return value here?

File: Detectors/TRD/qc/include/TRDQC/Tracking.h (hunk 4 above)
Comment (bazinski) [reply Detectors/TRD/qc/include/TRDQC/Tracking.h:91]: I dont remember hence my query.

File: Detectors/TRD/qc/include/TRDQC/Tracking.h (hunk 5 above)
Comment (f3sch) [reply Detectors/TRD/qc/include/TRDQC/Tracking.h]: Ok

File: Detectors/TRD/base/include/TRDBase/PadParameters.h (hunk 6 above)
Comment (f3sch) [reply Detectors/TRD/base/include/TRDBase/PadParameters.h]: dropped

**Review by martenole — DISMISSED** [review]
Thanks. It would of course also work the way you had it before, but I would like to keep all the framework stuff out of the detector code where it is not needed.

**Review by martenole — APPROVED** [review]
